pkg/controller/controller.go

```diff
@@ -1267,6 +1289,113 @@ func (c *M3DBController) getParentCluster(pod *corev1.Pod) (*myspec.M3DBCluster,
 	return cluster, nil
 }
 
+// cleanupUnusedStorageForSts removes all PVCs and PVs for the stateful which are not used.
+// This method should be called only when the cluster is stable
+// and there are no ongoing deployments.
+func (c *M3DBController) cleanupUnusedStorageForSts(
+	ctx context.Context,
+	cluster *myspec.M3DBCluster,
+	sts *appsv1.StatefulSet,
+) error {
+	logger := c.logger.With(
+		zap.String("cluster", cluster.Name),
+		zap.String("namespace", cluster.Namespace),
+	)
+
+	lbls := labels.BaseLabels(cluster)
+	lbls[labels.StatefulSet] = sts.Name
+
+	labelSelector := klabels.SelectorFromSet(lbls)
+	pvcs, err := c.pvcLister.PersistentVolumeClaims(cluster.Namespace).List(labelSelector)
+	if err != nil {
+		return fmt.Errorf("listing pvcs: %w", err)
+	}
+
+	pvcsToRemoveByName := make(map[string]*corev1.PersistentVolumeClaim, len(pvcs))
+	for _, pvc := range pvcs {
+		pvcsToRemoveByName[pvc.Name] = pvc
+	}
+
+	// first reclaim the PVCs which should be in use deciding by the parsed pod id
+	// and the replica count of the STS
+	lastPodIDInSTS := int(*sts.Spec.Replicas - 1)
+	for _, pvc := range pvcsToRemoveByName {
+		pvcPodID, err := parsePodIDFromPVCName(pvc.Name)
+		if err != nil {
+			logger.Error("Cannot parse pod id from the PVC name", zap.Error(err))
+
+			// reclaim the PVC just to be safe to not accidentally delete a PVC used by other applications
+			delete(pvcsToRemoveByName, pvc.Name)
+
+			continue
+		}
+
+		// PVC in use by one of the pods from the STS
+		if pvcPodID <= lastPodIDInSTS {
+			delete(pvcsToRemoveByName, pvc.Name)
+		}
+	}
+
+	// check whether there are any actually running pods which use any of the PVCs
+	pods, err := c.podLister.Pods(cluster.Namespace).List(labelSelector)
+	if err != nil {
+		return fmt.Errorf("listing pods: %w", err)
+	}
+
+	for _, pod := range pods {
+		for i := range pod.Spec.Volumes {
+			claim := pod.Spec.Volumes[i].PersistentVolumeClaim
+			if claim != nil {
+				delete(pvcsToRemoveByName, claim.ClaimName)
+			}
+		}
+	}
+
+	// Deleting a PVC which PV has a reclaim policy of "Delete" will delete PV automatically,
+	// so we don't need to do anything regarding PV deletion.
+	for _, pvc := range pvcsToRemoveByName {
+		// using kube client directly, since this should be a very rare call
+		// and will not cause any significant load on the API server
+		pv, err := c.kubeClient.CoreV1().
+			PersistentVolumes().
+			Get(ctx, pvc.Spec.VolumeName, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("getting k8s pv: %w", err)
+		}
+
+		// Sometimes a PV can have a reclaim policy other than "Delete". Skip these PVCs,
+		// so we would not leave dangling PVs which are hard to select because of the missing labels.
+		if pv.Spec.PersistentVolumeReclaimPolicy != corev1.PersistentVolumeReclaimDelete {
+			logger.Warn(
```

Comment:
We have seen glitches when PVs are not reclaimed even if they are "Delete" and pvc gets deleted. Maybe we should do "DeleteIfExists" for PV regardless if delete cascades or not.